### PR TITLE
- Add Redis Handler

### PR DESCRIPTION
**Pull request by daycry**
- Add Redis Handler

### CODE REVIEW

File: src/Handlers/RedisHandler.php [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+{
+    /**
+     * The Query Builder for the Settings table.
+     */
+    private BaseRedisHandler $handler;
+
+    /**
+     * Array of contexts that have been stored.
+     *
+     * @var ?string[]
```
Comment (MGatner) [inline src/Handlers/RedisHandler.php:23]: This type is unclear to me. Does this mean `string[]|null` or `array<string|null>`?

File: src/Handlers/RedisHandler.php [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+     */
+    private $hydrated = [];
+
+    /**
+     * Array of contexts that have been stored.
+     */
```
Comment (MGatner) [inline src/Handlers/RedisHandler.php:29]: Paste mistake!
```suggestion
```

File: src/Handlers/RedisHandler.php [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    private Cache $config;
+
+    /**
+     * Stores the configured database table.
+     */
```
Comment (MGatner) [inline src/Handlers/RedisHandler.php:34]: Another?
```suggestion
```

**Review by MGatner — CHANGES_REQUESTED** [review]
This looks very good! A few suggestions. I don't have a Redis dev environment right now but the tests look good. Maybe @paulbalandan or @kenjis has some Redis access? If not I will try to get a container going this weekend.

File: src/Handlers/RedisHandler.php [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        return $this->hasStored($class, $property, $context);
+    }
+
+    /**
+     * Attempt to retrieve a value from the database.
+     * To boost performance, all of the values are
+     * read and stored the first call for each contexts
+     * and then retrieved from storage.
+     *
+     * @return mixed|null
```
Comment (kenjis) [inline src/Handlers/RedisHandler.php:58]: Please do not use @mixed as possible.
If it is really `@mixed`, `null` should be removed.

File: src/Handlers/RedisHandler.php [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+     */
+    public function get(string $class, string $property, ?string $context = null)
+    {
+        return $this->getStored($class, $property, $context);
+    }
+
+    /**
+     * Stores values into the database for later retrieval.
+     *
+     * @param mixed $value
```
Comment (kenjis) [inline src/Handlers/RedisHandler.php:68]: Please do not use `@mixed` as possible.

File: src/Handlers/RedisHandler.php (hunk 1 above)
Comment (paulbalandan) [reply src/Handlers/RedisHandler.php:23]: This should read as `string[]|null`. `array<string|null>` would be `(string|null)[]`

File: src/Handlers/BaseRedisHandler.php [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace CodeIgniter\Settings\Handlers;
+
+use CodeIgniter\Cache\Handlers\RedisHandler;
+use Config\Cache;
+
+/**
+ * Redis class extension
+ */
+class BaseRedisHandler extends RedisHandler
```
Comment (paulbalandan) [inline src/Handlers/BaseRedisHandler.php:11]: `Handlers\BaseRedisHandler` and `Handlers\RedisHandler`, at first, baffles me. Why are there two Redis handlers which do not inherit from the other? The `Base` prefix, and with `Abstract`, connotes that wherever it is affixed that is meant to be inherited from. But this is apparently not the case.

I'm not sure if the design is good.

File: src/Handlers/BaseRedisHandler.php [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+{
+    public function __construct(Cache $config)
+    {
+        parent::__construct($config);
+        $this->initialize();
+    }
+
+    public function getRedis()
+    {
+        return $this->redis;
```
Comment (paulbalandan) [inline src/Handlers/BaseRedisHandler.php:21]: Why do you need to expose the `$redis` property? Is this for testing only?

File: src/Handlers/RedisHandler.php [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+     * The Query Builder for the Settings table.
+     */
+    private BaseRedisHandler $handler;
+
+    /**
+     * Array of contexts that have been stored.
+     *
+     * @var ?string[]
+     */
+    private $hydrated = [];
```
Comment (paulbalandan) [inline src/Handlers/RedisHandler.php:25]:
```suggestion
    private ?array $hydrated = [];
```

Comment (paulbalandan) [reply src/Handlers/RedisHandler.php:25]: I think the `null` is not needed?

File: src/Handlers/RedisHandler.php [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+    {
+        // Check for completion
+        if (in_array($context, $this->hydrated, true)) {
+            return;
+        }
+
+        if ($context === null) {
+            $this->hydrated[] = null;
+            $this->getHydrate('*--null');
+        // $query = $this->builder->where('context', null);
```
Comment (paulbalandan) [inline src/Handlers/RedisHandler.php:131]: Remove commented code.

File: src/Handlers/RedisHandler.php [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+        } else {
+            // $query = $this->builder->where('context', $context);
```
Comment (paulbalandan) [inline src/Handlers/RedisHandler.php:133]: Remove.

File: src/Handlers/BaseRedisHandler.php (hunk 7 above)
Comment (daycry) [reply src/Handlers/BaseRedisHandler.php:21]: No, I use redis property in getHydrate function to search with pattern.

$keys = $this->handler->getRedis()->scan($iterator, $pattern);

File: src/Handlers/RedisHandler.php (hunk 2 above)
Comment (daycry) [reply src/Handlers/RedisHandler.php:29]: I copied from DatabaseHandler.php

File: src/Handlers/RedisHandler.php (hunk 1 above)
Comment (daycry) [reply src/Handlers/RedisHandler.php:23]: I copied from DatabaseHandler.php

File: src/Handlers/BaseRedisHandler.php (hunk 6 above)
Comment (MGatner) [reply src/Handlers/BaseRedisHandler.php:11]: If we want to get rid of the class you could use a property reader. None of this is great practice but I totally get the intent. 
```php
$reader      = fn (): Redis => $this->redis;
$cache       = (new RedisHandler())->initialize();
$connection  = $reader->call($cache);
```

File: src/Handlers/RedisHandler.php (hunk 8 above)
Comment (MGatner) [reply src/Handlers/RedisHandler.php]: See my other comment; I think this is actually `array<string|null>`. I'm not sure if the shorthand is incorrect or if I just don't know about it 😂

File: src/Handlers/RedisHandler.php (hunk 1 above)
Comment (MGatner) [reply src/Handlers/RedisHandler.php]: 🤦‍♂️

File: src/Handlers/BaseRedisHandler.php (hunk 6 above)
Comment (paulbalandan) [reply src/Handlers/BaseRedisHandler.php:11]: It seems the use of this class is _just_ to retrieve the `Redis` instance. Why not use `Closure::bind()`?
```php
$cache = (new \CodeIgniter\Cache\Handlers\RedisHandler)->initialize();
$closure = Closure::bind(fn (): Redis => $this->redis, $cache, get_class($cache));

$closure();
```

Comment (kenjis) [reply src/Handlers/BaseRedisHandler.php:11]: Why Settings depends on Cache?
It seems bad practice.

Comment (paulbalandan) [reply src/Handlers/BaseRedisHandler.php:11]: > Why Settings depends on Cache? It seems bad practice.

The Cache dependency was just to retrieve the `Redis` instance.

Comment (kenjis) [reply src/Handlers/BaseRedisHandler.php:11]: Settings should not depend on Cache.
If we need a Redis instance, we can create it with `new Redis()`.

**Comment by lonnieezell** [comment]
It's been a very long time. Guessing this one's not getting finished. Closing.

**Pull request closed without merging**